openmc/model/triso.py

```diff
@@ -0,0 +1,172 @@
+import openmc.checkvalue as cv
+
+class TRISO(object):
+    """Tristructural-isotopic (TRISO) micro fuel particle
+
+    Parameters
+    ----------
+    materials : Iterable of openmc.Material
+        Material to be assigned to each layer of the TRISO particle starting
+        with the innermost and proceeding outwards
+    radii : Iterable of float
+        Outer radii in cm of each layer of the TRISO particle in ascending order
+    center : Iterable of float
+        Cartesian coordinates of the center of the TRISO particle in cm
+
+    Attributes
+    ----------
+    cells : list of opemc.Cell
+        Each layer of the TRISO particle
+    center : numpy.ndarray
+        Cartesian coordinates of the center of the TRISO particle in cm
+    outside : openmc.Region
+        Region of space outside of the TRISO particle
+    bounding_box : tuple of numpy.ndarray
+        Lower-left and upper-right coordinates of an axis-aligned bounding box
+        for the TRISO particle
+
+    """
+
+    def __init__(self, materials, radii, center=(0., 0., 0.)):
+        surfaces = [openmc.Sphere(R=r) for r in radii]
+        cells = []
+        for i, m in enumerate(materials):
+            c = openmc.Cell(fill=m)
+            if i == 0:
+                c.region = -surfaces[i]
+            else:
+                c.region = +surfaces[i-1] & -surfaces[i]
+            cells.append(c)
+        self._cells = cells
+        self._surfaces = surfaces
+        self.center = np.asarray(center)
+
+    @property
+    def bounding_box(self):
+        return self.cells[-1].region.bounding_box
+
+    @property
+    def cells(self):
+        return self._cells
+
+    @property
+    def center(self):
+        return self._center
+
+    @property
+    def outside(self):
+        return ~self.cells[-1].region.nodes[-1]
+
+    @center.setter
+    def center(self, center):
+        cv.check_type('TRISO center', center, Iterable, Real)
+        for s in self._surfaces:
+            s.x0, s.y0, s.z0 = center
+        self._center = center
+
+    def classify(self, lattice):
+        """Determine lattice element indices which might contain the TRISO particle.
+
+        Parameters
+        ----------
+        lattice : openmc.RectLattice
+            Lattice to check
+
+        Returns
+        -------
+        list of tuple
+            (z,y,x) lattice element indices which might contain the TRISO
+            particle.
+
+        """
+
+        ll, ur = self.bounding_box
+        if lattice.ndim == 2:
+            (i_min, j_min), p = lattice.find_element(ll)
+            (i_max, j_max), p = lattice.find_element(ur)
+            return list(np.broadcast(*np.ogrid[
+                j_min:j_max+1, i_min:i_max+1]))
+        else:
+            (i_min, j_min, k_min), p = lattice.find_element(ll)
+            (i_max, j_max, k_max), p = lattice.find_element(ur)
+            return list(np.broadcast(*np.ogrid[
+                k_min:k_max+1, j_min:j_max+1, i_min:i_max+1]))
```

Comment:
There should be a check here to make sure that all the `*_min` and `*_max` indices are valid.  Unfortunately, we can't use `lattice.is_valid_index` because `lattice._universes` isn't allocated yet